packages/lib-runtime/src/store/coderefs.ts

```diff
@@ -0,0 +1,150 @@
+import type { AnyObject } from '@monorepo/common';
+import { CustomError, visitDeep } from '@monorepo/common';
+import * as _ from 'lodash-es';
+import type { EncodedCodeRef, CodeRef, LoadedExtension } from '../types/extension';
+import type { PluginEntryModule } from '../types/runtime';
+
+class CodeRefError extends CustomError {
+  constructor(message: string, readonly cause?: unknown) {
+    super(message);
+  }
+}
+
+const codeRefSymbol = Symbol('CodeRef');
+
+/**
+ * Check if the given object has the {@link EncodedCodeRef} shape.
+ */
+const isEncodedCodeRef = (obj: unknown): obj is EncodedCodeRef =>
+  _.isPlainObject(obj) &&
+  _.isEqual(Object.getOwnPropertyNames(obj), ['$codeRef']) &&
+  typeof (obj as EncodedCodeRef).$codeRef === 'string';
+
+/**
+ * Check if the given object is a {@link CodeRef} function.
+ */
+const isCodeRef = (obj: unknown): obj is CodeRef =>
+  typeof obj === 'function' &&
+  _.isEqual(Object.getOwnPropertySymbols(obj), [codeRefSymbol]) &&
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (obj as any)[codeRefSymbol] === true;
+
+/**
+ * Parse data from the {@link EncodedCodeRef} object.
+ *
+ * Returns `undefined` if the `$codeRef` value is malformed.
+ */
+const parseEncodedCodeRef = (
+  ref: EncodedCodeRef,
+): [moduleName: string, exportName: string] | undefined => {
+  const match = ref.$codeRef.match(/^([^.]+)(?:\.([^.]+)){0,1}$/);
```

Comment:
FWIW, I think this also works:
```suggestion
  const match = ref.$codeRef.match(/^([^.]+)(?:\.([^.]+))?$/);
```